File: etap_reader/report_headers.py
"""Allowlisted presentation fields; never modify the uploaded study."""
import re

# id, label, source table, source column, maximum custom-text length
FIELDS = (
    ("sn", "SN", "Headr", "SN", 60),
    ("date", "Date", "Headr", "Date", 32),
    ("revision", "Revision", "ISCStudyCase", "Revision", 32),
    ("project", "Project", "Headr", "Project", 120),
    ("location", "Location", "Headr", "Loc", 120),
    ("contract", "Contract", "Headr", "Contr", 100),
    ("engineer", "Engineer", "Headr", "Eng", 100),
    ("filename", "Filename", "Headr", "FileN", 120),
    ("study_case", "Study case", "Headr", "STDCase", 120),
    ("configuration", "Configuration", "ISCStudyCase", "Config", 60),
    ("title_1", "Title line 1", "Headr", "1st", 180),
    ("title_2", "Title line 2", "Headr", "2nd", 180),
)
DEFINITIONS = {field[0]: field for field in FIELDS}


def validate(value):
    if not isinstance(value, dict) or any(key not in DEFINITIONS for key in value):
        raise ValueError("Header settings must contain only supported report fields.")
    result = {}
    for key, text in value.items():
        _, label, _, _, maximum = DEFINITIONS[key]
        if text is not None and (not isinstance(text, str) or len(text.encode('utf-16-le', errors='surrogatepass')) // 2 > maximum
                                 or re.search(r"[\x00-\x1f\x7f\ud800-\udfff]", text)):
            raise ValueError(f"{label} must be a single line of at most {maximum} characters, or hidden.")
        result[key] = text  # Absent = original, null = hide label and value, string = literal custom text.
    return result


def read_values(connection):
    result = {}
    tables = {row[0].lower(): row[0] for row in connection.execute("SELECT name FROM sqlite_master WHERE type='table'")}
    for key, _, table, column, _ in FIELDS:
        actual = tables.get(table.lower())
        if not actual:
            result[key] = ""
            continue
        quoted = '"' + actual.replace('"', '""') + '"'
        columns = {row[1].lower(): row[1] for row in connection.execute('PRAGMA table_info(' + quoted + ')')}
        actual_column = columns.get(column.lower())
        if not actual_column:
            result[key] = ""
            continue
        selected = '"' + actual_column.replace('"', '""') + '"'
        row = connection.execute('SELECT ' + selected + ' FROM ' + quoted + ' LIMIT 1').fetchone()
        result[key] = str(row[0])[:2000] if row and row[0] is not None else ""
    return result


def public_fields():
    return [{"id": key, "label": label, "max_length": maximum} for key, label, _, _, maximum in FIELDS]

File: desktop/EtapCrystalReporter/Services/ReportHeaders.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.IO;
using System.Linq;
using System.Text.RegularExpressions;
using EtapCrystalReporter.Models;

namespace EtapCrystalReporter.Services
{
    public static class ReportHeaders
    {
        // These are display fields only. In particular, Revision is the study
        // revision, not Headr.PSRev (the ETAP software version).
        private static readonly Dictionary<string, string[]> Fields = new Dictionary<string, string[]> {
            { "sn", new[] { "Headr", "SN", "60" } },
            { "date", new[] { "Headr", "Date", "32" } },
            { "revision", new[] { "ISCStudyCase", "Revision", "32" } },
            { "project", new[] { "Headr", "Project", "120" } },
            { "location", new[] { "Headr", "Loc", "120" } },
            { "contract", new[] { "Headr", "Contr", "100" } },
            { "engineer", new[] { "Headr", "Eng", "100" } },
            { "filename", new[] { "Headr", "FileN", "120" } },
            { "study_case", new[] { "Headr", "STDCase", "120" } },
            { "configuration", new[] { "ISCStudyCase", "Config", "60" } },
            { "title_1", new[] { "Headr", "1st", "180" } },
            { "title_2", new[] { "Headr", "2nd", "180" } }
        };

        public static void Validate(IDictionary<string, string> settings)
        {
            if (settings == null) return;
            foreach (var pair in settings)
            {
                if (!Fields.ContainsKey(pair.Key)) throw new InvalidDataException("Unsupported report header: " + pair.Key);
                string value = pair.Value;
                if (value != null && (value.Length > int.Parse(Fields[pair.Key][2]) || value.Any(char.IsControl)))
                    throw new InvalidDataException("Report header text is too long or contains a line break/control character: " + pair.Key);
            }
        }

        public static void ApplyToTable(DataTable table, IDictionary<string, string> settings)
        {
            Validate(settings);
            if (settings == null) return;
            foreach (var pair in settings)
            {
                // Only the two plain title fields require data binding. Named
                // header text is replaced in the report objects below, so study
                // case settings and database join fields retain their values.
                if (pair.Key != "title_1" && pair.Key != "title_2") continue;
                var field = Fields[pair.Key];
                if (!table.TableName.Equals(field[0], StringComparison.OrdinalIgnoreCase)) continue;
                var column = table.Columns.Cast<DataColumn>().FirstOrDefault(c => c.ColumnName.Equals(field[1], StringComparison.OrdinalIgnoreCase));
                if (column == null || column.DataType != typeof(string))
                    throw new InvalidDataException("This study does not contain the expected text header: " + pair.Key);
                column.MaxLength = -1;
                foreach (DataRow row in table.Rows) row[column] = pair.Value ?? "";
            }
        }

        public static void ApplyPresentation(dynamic report, string scope, DatabaseSnapshot database,
                                           ReportTemplate template, IDictionary<string, string> settings)
        {
            if (settings == null || settings.Count == 0) return;
            var replacements = new Dictionary<string, string>();
            foreach (dynamic table in report.Database.Tables)
            {
                string alias = (string)table.Name;
                string actual = TableBinding.Resolve(alias, (string)table.Location, scope, template, database.Info.Tables);
                foreach (var pair in settings)
                {
                    var field = Fields[pair.Key];
                    if (actual.Equals(field[0], StringComparison.OrdinalIgnoreCase)) replacements["{" + alias + "." + field[1] + "}"] = pair.Value;
                }
            }
            foreach (dynamic section in report.ReportDefinition.Sections)
            foreach (dynamic item in section.ReportObjects)
            {
                string kind = item.Kind.ToString();
                string text = kind == "TextObject" ? (string)item.Text : kind == "FieldObject" ? (string)item.DataSource.FormulaName : "";
                var matches = replacements.Where(pair => text.IndexOf(pair.Key, StringComparison.OrdinalIgnoreCase) >= 0).ToList();
                if (!matches.Any()) continue;
                if (matches.Any(pair => pair.Value == null))
                {
                    // ETAP embeds label and field in one text object. Removing
                    // the complete object hides both on every page.
                    if (kind == "TextObject") item.Text = "";
                    item.ObjectFormat.EnableSuppress = true;
                }
                else if (kind == "TextObject")
                {
                    foreach (var pair in matches)
                        text = Regex.Replace(text, Regex.Escape(pair.Key), match => pair.Value, RegexOptions.IgnoreCase);
                    item.Text = text; // Literal text, never an executable Crystal formula.
                }
            }
        }
    }
}
